python/mxnet/gluon/data/dataset.py

```diff
@@ -40,6 +40,41 @@ def __getitem__(self, idx):
     def __len__(self):
         raise NotImplementedError
 
+    def filter(self, fn):
```

Comment:
This will eagerly retrieve all values from a `lazy=True` transformed dataset. That should be documented?
(The transformation function will also be applied again in `__getitem__`)

Reply:
Yes I'll document that

Reply:
Is it better in some cases that lazy evaluation should be applied after `filter`, i.e., 

data.filter(xxx).transform is way better than data.transform.filter in some situations.

Reply:
Both are valid operations and I don't think we can stop users from doing transform() before filter(), since sometimes we need to actually filter the data based on the transformed data. For now I'd resort to better documentation